dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceOutputStream.java

```diff
@@ -222,79 +281,167 @@ private void makeInitialWriteRequest() throws Exception
         requestBuffer.reset();
         requestBuffer.putInt(packetLen);
 
-        this.socket.getOutputStream().write(requestBuffer.array(), 0, headerLen);
-        this.socket.getOutputStream().flush();
-        this.readResponse();
+        this.dos.write(requestBuffer.array(), 0, headerLen);
+        this.dos.flush();
+
+        RowServiceResponse response = this.readResponse();
+        this.handle = response.handle;
     }
 
-    /**
-     * Read response.
-     *
-     * @throws IOException
-     *             Signals that an I/O exception has occurred.
-     */
-    private void readResponse() throws IOException
+    private String makeCloseHandleRequest()
     {
-        this.scratchBuffer.clear();
-        this.socket.getInputStream().read(scratchBuffer.array(), 0, 4);
-        int len = scratchBuffer.getInt() & 0x7FFFFFFF;
+        StringBuilder sb = new StringBuilder(256);
+        sb.delete(0, sb.length());
+
+        sb.append("{ \"format\" : \"binary\",\n");
+        sb.append("  \"handle\" : \"" + Long.toString(this.handle) + "\",");
+        sb.append("  \"command\" : \"close\"");
+        sb.append("\n}");
+
+        return sb.toString();
+    }
+
+    private void sendCloseFileRequest() throws IOException
+    {
+        if (useOldProtocol)
+        {
+            return;
+        }
 
-        // We should always have a status & handle returned by the row service
-        if (len < 8)
+        String closeFileRequest = makeCloseHandleRequest();
+        int jsonRequestLen = closeFileRequest.length();
+
+        try
         {
-            throw new IOException("Received short or truncated response from row service. Aborting.");
+            this.dos.writeInt(jsonRequestLen + 4 + 1);
+            this.dos.write((int)RFCCodes.RFCStreamGeneral);
+            this.dos.writeInt(jsonRequestLen);
+            this.dos.write(closeFileRequest.getBytes(StandardCharsets.ISO_8859_1));
+            this.dos.flush();
+            System.out.println("Sent close request for handle: " + this.handle);
```

Comment:
Do we have a logger available?

Reply:
We do, but that is left over from debugging. Will remove that, good catch